src/ripple/app/tx/impl/details/NFTokenUtils.cpp

```diff
@@ -0,0 +1,498 @@
+    // NFTs must be kept on the same page.  So when the page contains
+    // equivalent NFTs, the split may be lopsided in order to keep equivalent
+    // NFTs on the same page.
+    STArray carr;
+    {
+        // We prefer to keep equivalent NFTs on a page boundary.  That gives
+        // any additional equivalent NFTs maximum room for expansion.
+        // Round up the boundary until there's a non-equivalent entry.
+        uint256 const cmp =
+            narr[(dirMaxTokensPerPage / 2) - 1].getFieldH256(sfTokenID) &
+            nft::pageMask;
+
+        // Note that the calls to find_if_not() and (later) find_if()
+        // rely on the fact that narr is kept in sorted order.
+        auto splitIter = std::find_if_not(
+            narr.begin() + (dirMaxTokensPerPage / 2),
+            narr.end(),
+            [&cmp](STObject const& obj) {
+                return (obj.getFieldH256(sfTokenID) & nft::pageMask) == cmp;
+            });
+
+        // If we get all the way from the middle to the end with only
+        // equivalent NFTokens then check the front of the page for a
+        // place to make the split.
+        if (splitIter == narr.end())
+            splitIter = std::find_if(
+                narr.begin(), narr.end(), [&cmp](STObject const& obj) {
+                    return (obj.getFieldH256(sfTokenID) & nft::pageMask) == cmp;
+                });
+
+        // If splitIter == begin(), then the entire page is filled with
+        // equivalent tokens.  We cannot split the page, so we cannot
+        // insert the requested token.
+        //
+        // There should be no circumstance when splitIter == end(), but if it
+        // were to happen we should bail out because something is confused.
+        if (splitIter == narr.begin() || splitIter == narr.end())
+            return nullptr;
+
+        // Split narr at splitIter.
+        STArray newCarr(
+            std::make_move_iterator(splitIter),
+            std::make_move_iterator(narr.end()));
+        narr.erase(splitIter, narr.end());
+        std::swap(carr, newCarr);
+    }
+
+    auto np = std::make_shared<SLE>(
+        keylet::nftpage(base, carr[0].getFieldH256(sfTokenID)));
+    np->setFieldArray(sfNonFungibleTokens, narr);
+    np->setFieldH256(sfNextPageMin, cp->key());
+
+    if (auto ppm = (*cp)[~sfPreviousPageMin])
+    {
+        np->setFieldH256(sfPreviousPageMin, *ppm);
+
+        if (auto p3 = view.peek(Keylet(ltNFTOKEN_PAGE, *ppm)))
+        {
+            p3->setFieldH256(sfNextPageMin, np->key());
+            view.update(p3);
+        }
+    }
+
+    view.insert(np);
+
+    cp->setFieldArray(sfNonFungibleTokens, carr);
+    cp->setFieldH256(sfPreviousPageMin, np->key());
+    view.update(cp);
+
+    createCallback(view, owner);
+
+    return (first.key <= np->key()) ? np : cp;
+}
+
+static bool
+compareTokens(uint256 const& a, uint256 const& b)
+{
+    return (a & nft::pageMask) < (b & nft::pageMask);
```

Comment:
I think there's a latent bug here.  We need every server to produce identically sorted `NFTPage`s.  But this comparison only sorts on the low 96-bits.  If there are NFTs with identical low 96-bits then the sort order becomes nondeterministic.

So I think this function needs to look more like this:
```
    // The sort of NFTokens needs to be fully deterministic, but the sort
    // is weird because we sort on the low 96-bits first.  But if the low
    // 96-bits are identical we still need a fully deterministic sort.
    // So we sort on the low 96-bits first.  If those are equal we sort on
    // the whole thing.
    int const lowBitsCmp = compare(a & nft::pageMask, b & nft::pageMask);
    if (lowBitsCmp != 0)
        return lowBitsCmp < 0;

    return a < b;
```